runtime/gc/gc.go

```diff
@@ -61,11 +93,59 @@ func (gc *GarbageCollector) Unregister(resource Disposable) bool {
 	return false
 }
 
-// Collect runs garbage collection and reclaims resources, at this stage it just
-// calls CollectAll(), but in the future this should dispose resources somewhat
-// intelligently.
+// Collect runs garbage collection and reclaims resources, attempting to
+// satisfy minimumMemory and minimumDiskSpace, if possible.
 func (gc *GarbageCollector) Collect() error {
-	return gc.CollectAll()
+	gc.m.Lock()
+	defer gc.m.Unlock()
+
+	// Sort to get least-recently-used first
+	sort.Sort(disposableSorter(gc.resources))
+
+	var resources []Disposable
+	for i, r := range gc.resources {
+		var err error
+		var size uint64
+
+		if gc.needDiskSpace() {
+			size, err = r.DiskSize()
+			if err != nil && err != ErrDisposableSizeNotSupported {
+				goto abort
+			}
+			if size > 0 || err == ErrDisposableSizeNotSupported {
+				goto dispose
+			}
+		}
+
+		if gc.needMemory() {
+			size, err = r.MemorySize()
+			if err != nil && err != ErrDisposableSizeNotSupported {
+				goto abort
+			}
+			if size > 0 || err == ErrDisposableSizeNotSupported {
+				goto dispose
+			}
+		}
+
+		goto keep
+	keep:
+		resources = append(resources, r)
+		continue
+	abort:
+		gc.resources = append(resources, gc.resources[i-1:]...)
+		return err
+	dispose:
+		err = r.Dispose()
+		if err != nil {
+			if err != ErrDisposableInUse {
+				goto abort
+			}
+			goto keep
+		}
+	}
```

Comment:
Yeah, I think this is something we'll refactor later... We should possibly use the estimate resource sizes to decide how many we release... Things like that...